rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em

```diff
@@ -459,6 +462,288 @@ if isinstance(type_, AbstractNestedType):
 }
 
+@[  if message.structure.has_any_member_with_annotation('key') ]@
+static
+size_t
+_@(message.structure.namespaced_type.name)__max_serialized_key_size(
+  size_t initial_alignment,
+  bool & is_unbounded)
+{
+  size_t current_alignment = initial_alignment;
+  is_unbounded = false;
+@[for member in message.structure.members]@
+@[  if not member.has_annotation('key')]@
+@[  continue]@
+@[  end if]@
+  // Member: @(member.name)
+  {
+@[    if isinstance(member.type, AbstractNestedType) ]@
+@[    if isinstance(member.type, Array)]@
+    size_t array_size = @(member.type.size);
+@[    elif isinstance(member.type, BoundedSequence)]@
+    size_t array_size = @(member.type.maximum_size);
+@[    else]@
+    size_t array_size = 0;
+@[    end if]@
+@[    if isinstance(member.type, AbstractSequence)]@
+    const size_t padding = 4;
+    current_alignment += padding +
+      eprosima::fastcdr::Cdr::alignment(current_alignment, padding);
+@[    end if]@
+@[  else]@
+    size_t array_size = 1;
+@[  end if]@
+
+@{
+type_ = member.type
+if isinstance(type_, AbstractNestedType):
+    type_ = type_.value_type
+}@
+@[  if isinstance(type_, AbstractGenericString)]@
+    is_unbounded = true;
+    const size_t padding = 4;
+    const size_t wchar_size = 4;
+
+    for (size_t index = 0; index < array_size; ++index) {
+      current_alignment += padding +
+        eprosima::fastcdr::Cdr::alignment(current_alignment, padding) +
+@[    if type_.has_maximum_size()]@
+@[      if isinstance(type_, AbstractWString)]@
+        wchar_size *
+@[      end if]@
+        @(type_.maximum_size) +
+@[    end if]@
+@[    if isinstance(type_, AbstractWString)]@
+        wchar_size *
+@[    end if]@
+        1;
+    }
+@[  elif isinstance(type_, BasicType)]@
+@[    if type_.typename in ('boolean', 'octet', 'char', 'uint8', 'int8')]@
+    current_alignment += array_size * sizeof(uint8_t);
+@[    elif type_.typename in ('wchar', 'int16', 'uint16')]@
+    current_alignment += array_size * sizeof(uint16_t) +
+      eprosima::fastcdr::Cdr::alignment(current_alignment, sizeof(uint16_t));
+@[    elif type_.typename in ('int32', 'uint32', 'float')]@
+    current_alignment += array_size * sizeof(uint32_t) +
+      eprosima::fastcdr::Cdr::alignment(current_alignment, sizeof(uint32_t));
+@[    elif type_.typename in ('int64', 'uint64', 'double')]@
+    current_alignment += array_size * sizeof(uint64_t) +
+      eprosima::fastcdr::Cdr::alignment(current_alignment, sizeof(uint64_t));
+@[    elif type_.typename == 'long double']@
+    current_alignment += array_size * sizeof(long double) +
+      eprosima::fastcdr::Cdr::alignment(current_alignment, sizeof(long double));
+@[    end if]@
+@[  else]
+    for (size_t index = 0; index < array_size; ++index) {
+      current_alignment +=
+        @('::'.join(type_.namespaces))::typesupport_fastrtps_cpp::_@(message.structure.namespaced_type.name)__max_serialized_key_size(
+        current_alignment, is_unbounded);
```

Comment:
We might have a bug here. If the nested type is the last one in the message, it could incorrectly override `is_unbounded`.

Consider the following types:

```
struct Inner
{
   int a;
   int b;
};

struct Top
{
    string id;
    Inner data;
};
```

In this case, Top should be unbounded